text/0162-refactoring-support.md

```diff
@@ -0,0 +1,622 @@
+customers as a workaround, adding unnecessary cognitive load for developers.
+Code refactoring is a fundamental job-to-be-done that developers expect to be
+supported natively by the tool.
+
+In addition to this, the recent launch of CloudFormation's stack refactoring API
+made it possible to support refactoring on the service side. We are building on
+top of that API to bring a seamless experience to CDK users.
+
+### Why should we _not_ do this?
+
+The main attraction of this feature is also, in a way, its greatest deterrent:
+that the refactoring happens automatically, and that the CLI makes decisions
+that the user may not even be aware they need to make (accidental renames, for
+example). This may cause anxiety for some users, who might not understand what
+exactly is happening, and what the consequences are. We can mitigate this risk
+with good documentation and careful interaction design.
+
+### What is the technical solution (design) of this feature?
+
+High level description of the algorithm:
+
+First, list all the stacks: both local and deployed. Then build an index of all
+resources from all stacks. This index maps the _content_ address (physical ID or
+digest) of each resource to all the _location_ addresses (stack name + logical
+ID) they can be found in.
+
+Resources that have different locations before and after, are considered to have
+been moved. For each of those, create a mapping from the source (the currently
+deployed location) to the destination (the new location, in the local template).
+Example:
+
+    // Keys are the content address of the resources
+    // Values are the location addresses
+    { 
+      "5e19886121239b7a": { // Moved and renamed -> goes to mapping
+        "before": ["stack1/logicalId1"],
+        "after": "["stack2/logicalId2"]
+      },    
+      "24ad8195002086b6": { // Removed -> ignored 
+        "before": ["stack1/logicalId3"]
+      },
+      "07266c4dd0146e8a": { // Unchanged -> ignored 
+        "before": ["stack1/logicalId4"]
+        "after": ["stack1/logicalId4"]
+      }
+    }
+
+Since the CloudFormation API expects not only the mappings, but also the
+templates in their final states, we need to compute those as well. This is done
+by applying all the mappings locally, essentially emulating what CloudFormation
+will eventually do. For example, if a mapping says that a resource has moved
+from stack `A` with name `Foo`, to stack `B` with name `Bar`, we will remove
+`Foo` from the template for stack `A`, and add a new resource called `Bar` to
+the template for stack `B`.
+
+At this point, if there is any ambiguity (more than one source or more than one
+destination for a given mapping), stop and ask the user whether they want to
+proceed.
+
+Assuming there were no ambiguities, or the user wants to proceed anyway, it is
+ready to call the API to actually perform the refactor, using the mappings and
+templates computed previously.
+
+### Is this a breaking change?
+
+No.
```

Comment:
Call-out: We need to implement this carefully for `cdk deploy` otherwise it could accidentally break Cx.